Mathlib/Analysis/Calculus/Gradient/Basic.lean

```diff
@@ -0,0 +1,370 @@
+and `gradf` be a vector in F. Then
+
+  `HasGradientWithinAt f f' s x`
+
+says that `f` has a gradient `f'` at `x`, where the domain of interest
+is restricted to `s`. We also have
+
+  `HasGradientAt f f' x := HasGradientWithinAt f f' x univ`
+
+Finally,
+
+  `HasGradientAt f f' x`
+
+means that `f : F → 𝕜` has a gradient `f' : F` in the sense of strict differentiability,
+i.e., `f y - f z - ⟨ f' , y - z ⟩ = o(y - z)` as `y, z → x`.
+
+## Main results
+
+This file contains the following parts of gradient.
+* the definition of gradient.
+* the theorems translating between `HasGradientAtFilter` and `HasFDerivAtFilter`,
+  `HasGradientWithinAt` and `HasFDerivWithinAt`, `HasGradientAt` and `HasFDerivAt`,
+  `Gradient` and `fderiv`.
+* theorems the Uniqueness of Gradient.
+* the theorems translating between  `HasGradientAtFilter` and `HasDerivAtFilter`,
+  `HasGradientAt` and `HasDerivAt`, `Gradient` and `deriv` when `F = 𝕜`.
+* the theorems about the congruence of the gradient.
+* the theorems about the gradient of constant function.
+* the theorems about the continuity of a function admitting a gradient.
+-/
+
+open Topology InnerProductSpace Set
+
+noncomputable section
+
+variable {𝕜 F : Type*} [IsROrC 𝕜]
+
+variable [NormedAddCommGroup F] [InnerProductSpace 𝕜 F] [CompleteSpace F]
+
+variable {f : F → 𝕜} {f' x : F}
+
+/-- A function `f` has the gradient `f'` as derivative along the filter `L` if
+  `f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` when `x'` converges along the filter `L`.-/
+def HasGradientAtFilter (f : F → 𝕜) (f' x : F) (L : Filter F) :=
+  HasFDerivAtFilter f (toDual 𝕜 F f') x L
+
+/-- `f` has the gradient `f'` at the point `x` within the subset `s` if
+  `f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x` inside `s`.-/
+def HasGradientWithinAt (f : F → 𝕜) (f' : F) (s : Set F) (x : F):=
+  HasGradientAtFilter f f' x (𝓝[s] x)
+
+/-- `f` has the gradient `f'` at the point `x` if
+  `f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x`. -/
+def HasGradientAt (f : F → 𝕜) (f' x : F) :=
+  HasGradientAtFilter f f' x (𝓝 x)
+
+/-- Gradient of `f` at the point `x`, if it exists.  Zero otherwise.
+If the derivative exists (i.e., `∃ f', HasGradientAt f f' x`), then
+`f x' = f x + ⟨ f', x' - x ⟩ + o (x' - x)` where `x'` converges to `x`. -/
+def Gradient (f : F → 𝕜) (x : F) : F := (toDual 𝕜 F).symm (fderiv 𝕜 f x)
+
+local notation "∇" => Gradient
+
+variable {s : Set F} {L : Filter F}
+
+theorem HasGradientWithinAt_iff_HasFDerivWithinAt {s : Set F} :
+  HasGradientWithinAt f f' s x ↔ HasFDerivWithinAt f (toDual 𝕜 F f') s x := by
+  constructor <;> (intro h; exact h)
+
+theorem HasGradientWithinAt.hasFDerivWithinAt {s : Set F} (h : HasGradientWithinAt f f' s x) :
+  HasFDerivWithinAt f (toDual 𝕜 F f') s x := HasGradientWithinAt_iff_HasFDerivWithinAt.mp h
+
+theorem HasFDerivWithinAt.hasGradientWithinAt {s : Set F}
+  (h : HasFDerivWithinAt f (toDual 𝕜 F f') s x) :
```

Comment:
This theorem should rather read: if `HasFDerivWithinAt f u s x` then `HasGradientWithinAt f ((toDual 𝕜 F).symm u) s x`. And maybe add the iff version also.
Edit: I see that the iff version is below. Then I would just drop the current lemma (which is quite unnatural, if needed one can resort to the iff version) and only keep the natural version. And probably move the other versions very close to here, as they have exactly the same flavor.